console/src/state/resources.rs

```diff
@@ -0,0 +1,332 @@
+use crate::state::{truncate_registry_path, Field, Metadata};
+use crate::view;
+use console_api as proto;
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    convert::{TryFrom, TryInto},
+    rc::{Rc, Weak},
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+use tui::{
+    style::{Color, Modifier},
+    text::Span,
+};
+
+#[derive(Default, Debug)]
+pub(crate) struct ResourcesState {
+    resources: HashMap<u64, Rc<RefCell<Resource>>>,
+    new_resources: Vec<ResourceRef>,
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(usize)]
+pub(crate) enum SortBy {
+    Rid = 0,
+    Kind = 1,
+    ConcreteType = 2,
+    Target = 3,
+    Total = 4,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub(crate) enum Kind {
+    Timer,
+    Other(String),
+}
+
+#[derive(Debug)]
+pub(crate) struct Resource {
+    id: u64,
+    meta_id: u64,
+    kind: Kind,
+    stats: ResourceStats,
+    target: Arc<str>,
+    concrete_type: String,
+    location: String,
+}
+
+pub(crate) type ResourceRef = Weak<RefCell<Resource>>;
+
+#[derive(Debug)]
+struct ResourceStats {
+    created_at: SystemTime,
+    dropped_at: Option<SystemTime>,
+    total: Option<Duration>,
+    attributes: Vec<Attribute>,
+    formatted_attributes: Vec<Vec<Span<'static>>>,
+}
+
+#[derive(Debug)]
+pub(crate) struct Attribute {
+    field: Field,
+    unit: Option<String>,
+}
```

Comment:
style nit: mind putting the `pub(crate)` struct before the module-private one?
```suggestion
#[derive(Debug)]
pub(crate) struct Attribute {
    field: Field,
    unit: Option<String>,
}

#[derive(Debug)]
struct ResourceStats {
    created_at: SystemTime,
    dropped_at: Option<SystemTime>,
    total: Option<Duration>,
    attributes: Vec<Attribute>,
    formatted_attributes: Vec<Vec<Span<'static>>>,
}
```